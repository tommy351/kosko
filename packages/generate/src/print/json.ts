import stringify from "fast-safe-stringify";
import { Printer, Writer } from "./types";

const EOL = "\n";
const INDENT = "  ";

export class JSONPrinter implements Printer {
  private written = false;
  private buffer: unknown;

  public constructor(private writer: Writer) {}

  public print(value: unknown) {
    if (this.written) {
      this.writeValue(value);
    } else if (this.buffer != null) {
      this.written = true;

      this.writeListStart();
      this.writeValue(this.buffer);

      this.buffer = null;

      this.writeValue(value);
    } else {
      this.buffer = value;
    }
  }

  public flush() {
    if (this.buffer != null) {
      this.writeValue(this.buffer);
    }

    if (this.written) {
      this.writeListEnd();
    }
  }

  private writeValue(value: unknown) {
    let result = stringify(value, undefined, INDENT);

    if (this.written) {
      if (this.buffer == null) {
        this.writer.write(",\n");
      }

      result = result
        .split(EOL)
        .map((line) => `${INDENT}${INDENT}${line}`)
        .join(EOL);
    }

    this.writer.write(result);
  }

  private writeListStart() {
    this.writer.write(`{
  "apiVersion": "v1",
  "kind": "List",
  "items": [
`);
  }

  private writeListEnd() {
    this.writer.write(`
  ]
}`);
  }
}
