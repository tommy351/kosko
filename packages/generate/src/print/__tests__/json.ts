import BufferList from "bl";
import { JSONPrinter } from "../json";

test("flush immediately without calling print", () => {
  const writer = new BufferList();
  const printer = new JSONPrinter(writer);

  printer.flush();

  expect(writer.toString()).toEqual("");
});

test("print a single value", () => {
  const writer = new BufferList();
  const printer = new JSONPrinter(writer);
  const data = { foo: "bar" };

  printer.print(data);
  printer.flush();

  expect(JSON.parse(writer.toString())).toEqual(data);
  expect(writer.toString()).toMatchSnapshot();
});

test("print multiple values", () => {
  const writer = new BufferList();
  const printer = new JSONPrinter(writer);
  const data = [{ a: "b" }, { c: "d" }, { e: "f" }];

  for (const item of data) {
    printer.print(item);
  }

  printer.flush();

  expect(JSON.parse(writer.toString())).toEqual({
    apiVersion: "v1",
    kind: "List",
    items: data
  });
  expect(writer.toString()).toMatchSnapshot();
});
