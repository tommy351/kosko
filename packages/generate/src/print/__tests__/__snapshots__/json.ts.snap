// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`print a single value 1`] = `
"{
  "foo": "bar"
}"
`;

exports[`print multiple values 1`] = `
"{
  "apiVersion": "v1",
  "kind": "List",
  "items": [
    {
      "a": "b"
    },
    {
      "c": "d"
    },
    {
      "e": "f"
    }
  ]
}"
`;
